signatures/pdf_sign_external_globalsign.go

```diff
@@ -0,0 +1,586 @@
+		if err != nil {
+			return nil, err
+		}
+		return signature, nil
+	}
+
+	siConfig := pkcs7.SignerInfoConfig{}
+	if len(certs) > 1 {
+		// Verify OCSP response.
+		_, err := ocsp.ParseResponseForCert(es.ocsp, certs[0], certs[1])
+		if err != nil {
+			return err
+		}
+
+		siConfig.ExtraSignedAttributes = []pkcs7.Attribute{
+			{
+				Type: pkcs7.OIDAttributeAdobeRevocation,
+				Value: RevocationInfoArchival{
+					Crl: []asn1.RawValue{},
+					Ocsp: []asn1.RawValue{
+						{FullBytes: es.ocsp},
+					},
+					OtherRevInfo: []asn1.RawValue{},
+				},
+			},
+		}
+	}
+
+	// If contains certificate chains.
+	if len(certs) > 1 {
+		err = signedData.AddSignerChain(certs[0], NewSigner(cb), certs[1:], siConfig)
+		if err != nil {
+			return err
+		}
+	} else if len(certs) == 1 {
+		err = signedData.AddSigner(certs[0], NewSigner(cb), siConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Add timestamp token to first signer.
+	err = signedData.RequestSignerTimestampToken(0, func(digest []byte) ([]byte, error) {
+		hasher := sha256.New()
+		hasher.Write(digest)
+		hasher.Sum(nil)
+
+		// Request timestamp token.
+		t, err := gsClient.DSSService.DSSIdentityTimestamp(es.ctx, "UniDoc", &globalsign.IdentityRequest{SubjectDn: globalsign.SubjectDn{}}, hasher.Sum(nil))
+		if err != nil {
+			return nil, err
+		}
+
+		return t, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Call Detach() is you want to remove content from the signature
+	// and generate an S/MIME detached signature.
+	signedData.Detach()
+	// Finish() to obtain the signature bytes.
+	detachedSignature, err := signedData.Finish()
+	if err != nil {
+		return err
+	}
+
+	data := make([]byte, sigLen)
+	copy(data, detachedSignature)
+
+	sig.Contents = core.MakeHexString(string(data))
+	return nil
+}
+
+// IsApplicable .
+func (es *externalSigner) IsApplicable(sig *model.PdfSignature) bool {
+	if sig == nil || sig.Filter == nil || sig.SubFilter == nil {
+		return false
+	}
+
+	return (*sig.Filter == "Adobe.PPKMS" || *sig.Filter == "Adobe.PPKLite") && *sig.SubFilter == "adbe.pkcs7.detached"
+}
+
+// NewDigest creates a new digest.
+func (es *externalSigner) NewDigest(sig *model.PdfSignature) (model.Hasher, error) {
+	return bytes.NewBuffer(nil), nil
+}
+
+// Validate validates PdfSignature.
+func (es *externalSigner) Validate(sig *model.PdfSignature, digest model.Hasher) (model.SignatureValidationResult, error) {
+	return model.SignatureValidationResult{
+		IsSigned:   true,
+		IsVerified: true,
+	}, nil
+}
+
+// getCertificateChain returns certificate chain.
+func (es *externalSigner) getCertificateChain() []*x509.Certificate {
+	return es.certChain
+}
+
+// RevocationInfoArchival is OIDAttributeAdobeRevocation attribute.
+type RevocationInfoArchival struct {
+	Crl          []asn1.RawValue `asn1:"explicit,tag:0,optional"`
+	Ocsp         []asn1.RawValue `asn1:"explicit,tag:1,optional"`
+	OtherRevInfo []asn1.RawValue `asn1:"explicit,tag:2,optional"`
+}
+
+// SignerCallback .
```

Comment:
Please complete the comment.

Reply:
Updated at ddbee0a38a3f9965ffe9ba4a532f1aa855f95346